From bace336e239a081244788f5c4ec2f00443b0cf2e Mon Sep 17 00:00:00 2001
From: Foxes Hung <foxes687@andestech.com>
Date: Mon, 27 Nov 2023 10:01:47 +0800
Subject: [PATCH 124/241] semihosting:   Default to access semihosting console
 when user directly writes stdin/stdout/stderr even user doesn't open fd 0,1,2
   This is a workaround for bug #29062

Signed-off-by: Foxes Hung <foxes687@andestech.com>
---
 semihosting/syscalls.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/semihosting/syscalls.c b/semihosting/syscalls.c
index 7f704ac3a3..7f73ddfff8 100644
--- a/semihosting/syscalls.c
+++ b/semihosting/syscalls.c
@@ -16,6 +16,7 @@
 #include "qemu.h"
 #else
 #include "semihosting/uaccess.h"
+#include CONFIG_DEVICES
 #endif
 
 
@@ -752,6 +753,12 @@ void semihost_sys_read(CPUState *cs, gdb_syscall_complete_cb complete,
                        int fd, target_ulong buf, target_ulong len)
 {
     GuestFD *gf = get_guestfd(fd);
+#ifdef CONFIG_ARM_COMPATIBLE_SEMIHOSTING
+    /* Andes workaround for bug 29062, default to directly use stdin */
+    if (gf == NULL && fd == 0) {
+        gf = &console_in_gf;
+    }
+#endif
 
     if (gf) {
         semihost_sys_read_gf(cs, complete, gf, buf, len);
@@ -794,6 +801,14 @@ void semihost_sys_write(CPUState *cs, gdb_syscall_complete_cb complete,
                         int fd, target_ulong buf, target_ulong len)
 {
     GuestFD *gf = get_guestfd(fd);
+#ifdef CONFIG_ARM_COMPATIBLE_SEMIHOSTING
+    /* Andes workaround for bug 29062, default to directly use fd 1,2 */
+    if (gf == NULL) {
+        if (fd == 1 || fd == 2) {
+            gf = &console_out_gf;
+        }
+    }
+#endif
 
     if (gf) {
         semihost_sys_write_gf(cs, complete, gf, buf, len);
-- 
2.34.1

